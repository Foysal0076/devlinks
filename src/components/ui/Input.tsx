import clsx from 'clsx'
import React, { forwardRef } from 'react'

type TextFieldParams = React.InputHTMLAttributes<HTMLInputElement> & {
  label?: string
  helperText?: string
  layout?: 'horizontal' | 'vertical'
  error?: string
  startAdornment?: StartAdornment
  endAdornment?: EndAdornment
}

type StartAdornment = {
  adornment: React.ReactNode
  onClick?: () => void
  className?: string
}

type EndAdornment = {
  adornment: React.ReactNode
  onClick: () => void
  className?: string
}

export const Input = forwardRef<HTMLInputElement, TextFieldParams>(
  function CustomInput(
    {
      label,
      error,
      layout = 'vertical',
      className = '',
      helperText,
      startAdornment,
      endAdornment,
      ...inputParams
    },
    ref
  ) {
    return (
      <div
        className={clsx('custom-input group flex', {
          'flex-col gap-1': layout === 'vertical',
        })}>
        {label && (
          <label
            htmlFor={inputParams.id as string}
            className={clsx(
              'text-sm font-medium',
              {
                'text-neutral-600 group-focus-within:text-primary-500 dark:group-focus-within:text-primary-300':
                  !error,
              },
              { 'text-red-500 group-focus-within:text-red-500': error },
              {
                'after:ml-0.5 after:mr-1 after:text-red-500 after:content-["*"]':
                  inputParams.required,
              },
              {
                'mt-2': layout === 'horizontal',
              }
            )}>
            {label}
            {layout === 'horizontal' && <span>:</span>}
          </label>
        )}
        <div className='relative flex flex-col'>
          <input
            ref={ref}
            {...inputParams}
            className={clsx(
              `rounded-lg border-none px-4 text-neutral-600 outline-none ring-1 placeholder:text-neutral-100 focus:placeholder:text-primary-100 disabled:bg-neutral-30 disabled:text-neutral-100 dark:bg-neutral-800 dark:text-neutral-50 dark:disabled:bg-neutral-400 ${className}`,
              {
                'ring-neutral-50 focus:shadow-[0_0_10px_#c9bfff] focus:ring-primary-300 dark:focus:ring-primary-300':
                  !error,
                'ring-red-500 focus:ring-red-500': error,
                'pr-12': endAdornment,
                'pl-10': startAdornment,
              }
            )}
          />
          {helperText && (
            <span className='helper-text text-xs font-normal text-neutral-400 md:text-sm'>
              {`${helperText}`}
            </span>
          )}
          {error && (
            <div
              className={clsx(
                'absolute -bottom-0 left-0',
                { 'translate-y-[80%]': helperText },
                { 'translate-y-[90%]': !helperText }
              )}>
              <span className='-mt-1 text-xs text-red-500 dark:text-red-500'>
                {error}
              </span>
            </div>
          )}
          {startAdornment && (
            <button
              type='button'
              onClick={startAdornment?.onClick}
              className={clsx(
                'absolute left-2 top-1/2 w-max -translate-y-1/2 rounded-full text-neutral-50 transition-colors duration-200 hover:bg-neutral-30',
                `${startAdornment.className ? startAdornment.className : ''}`,
                { 'pointer-events-none': !startAdornment.onClick }
              )}
              aria-label='start-adornment btn'>
              {startAdornment.adornment}
            </button>
          )}
          {endAdornment && (
            <button
              type='button'
              onClick={endAdornment?.onClick}
              className={clsx(
                'absolute right-2 top-1/2 w-max -translate-y-1/2 rounded-full text-neutral-50 transition-colors duration-200 hover:bg-neutral-30',
                `${endAdornment.className ? endAdornment.className : ''}`,
                { 'pointer-events-none': !endAdornment.onClick }
              )}
              aria-label='end-adornment btn'>
              {endAdornment.adornment}
            </button>
          )}
        </div>
      </div>
    )
  }
)
