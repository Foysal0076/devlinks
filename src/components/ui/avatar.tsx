import Image from 'next/image'

import { cn,createInitial } from '@/shared/utils'

type Props = {
  name: string
  url?: string
  size?: number
  className?: string
  textClassName?: string
}

const Avatar = ({ name, url, className, textClassName }: Props) => {
  const initial = createInitial(name)

  return (
    <div
      className={cn(
        `relative flex h-8 w-8 items-center justify-center rounded-full border-neutral-100 bg-surface-300 text-base font-bold`,
        className
      )}>
      {url ? (
        <Image
          src={url}
          alt={name}
          fill
          className='rounded-full object-cover'
        />
      ) : (
        <span className={cn('text-sm font-bold md:text-base', textClassName)}>
          {initial}
        </span>
      )}
    </div>
  )
}

export default Avatar
