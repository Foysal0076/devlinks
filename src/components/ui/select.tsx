import { CaretDown, Check } from '@phosphor-icons/react'
import React, {
  forwardRef,
  memo,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react'

import { cn } from '@/shared/utils'

export type SelectOption = {
  value: string
  label: string
  icon?: React.ReactNode
}

type SelectProps = {
  options: SelectOption[]
  value: string
  onChange: (value: string) => void
  required?: boolean
  label?: string
  labelClassName?: string
  caretClassName?: string
  error?: boolean
  errorMessage?: string
  disabled?: boolean
}

// Memoized option component
const SelectItem = memo(
  ({
    option,
    isSelected,
    isHighlighted,
    onSelect,
  }: {
    option: SelectOption
    isSelected: boolean
    isHighlighted: boolean
    onSelect: () => void
  }) => (
    <div
      role='option'
      aria-selected={isSelected}
      onClick={onSelect}
      className={cn(
        'flex cursor-pointer items-center gap-2 px-3 py-2 hover:bg-neutral-50',
        isSelected ? `dark:bg-primary-0 bg-primary-50 text-primary-800` : '',
        isHighlighted ? 'bg-neutral-50' : ''
      )}>
      {option.icon && <>{option.icon}</>}
      <span className='block select-none truncate text-inherit'>
        {option.label}
      </span>
      {isSelected && <Check className='ml-auto h-4 w-4' />}
    </div>
  )
)

SelectItem.displayName = 'SelectItem'

const Select = forwardRef<HTMLDivElement, SelectProps>(
  (
    {
      options,
      value,
      onChange,
      label = 'Select Option',
      error = false,
      errorMessage,
      disabled = false,
      required,
      labelClassName = '',
      caretClassName = '',
    },
    ref
  ) => {
    const [isOpen, setIsOpen] = useState(false)
    const [highlightedIndex, setHighlightedIndex] = useState<number>(-1)
    const dropdownRef = useRef<HTMLDivElement>(null)
    const containerRef = useRef<HTMLDivElement>(null)

    const selectedOption = options.find((opt) => opt.value === value)
    const closeDropdown = useCallback(() => {
      setIsOpen(false)
      setHighlightedIndex(-1)
      containerRef.current?.focus()
    }, [])

    const handleSelect = useCallback(
      (optionValue: string) => {
        onChange(optionValue)
        closeDropdown()
      },
      [onChange, closeDropdown]
    )

    const toggleOpen = useCallback(() => {
      if (!disabled) {
        setIsOpen((prev) => !prev)
      }
    }, [disabled])

    useEffect(() => {
      const handleClickOutside = (event: MouseEvent) => {
        if (
          dropdownRef.current &&
          !dropdownRef.current.contains(event.target as Node) &&
          !containerRef.current?.contains(event.target as Node)
        ) {
          closeDropdown()
        }
      }

      document.addEventListener('mousedown', handleClickOutside)
      return () => document.removeEventListener('mousedown', handleClickOutside)
    }, [closeDropdown])

    const handleKeyDown = useCallback(
      (event: React.KeyboardEvent) => {
        if (disabled) return

        switch (event.key) {
          case 'Enter':
          case ' ':
            if (!isOpen) {
              setIsOpen(true)
            } else if (highlightedIndex !== -1) {
              handleSelect(options[highlightedIndex].value)
            }
            event.preventDefault()
            break
          case 'ArrowDown':
            event.preventDefault()
            if (!isOpen) {
              setIsOpen(true)
            } else {
              setHighlightedIndex((prev) =>
                prev === options.length - 1 ? 0 : prev + 1
              )
            }
            break
          case 'ArrowUp':
            event.preventDefault()
            if (!isOpen) {
              setIsOpen(true)
            } else {
              setHighlightedIndex((prev) =>
                prev <= 0 ? options.length - 1 : prev - 1
              )
            }
            break
          case 'Escape':
            closeDropdown()
            break
          case 'Tab':
            if (isOpen) {
              closeDropdown()
            }
            break
        }
      },
      [isOpen, highlightedIndex, options, handleSelect, disabled, closeDropdown]
    )

    return (
      <div ref={ref} className='relative w-full min-w-[200px]'>
        {label && (
          <label
            className={cn(
              'text-xs font-medium',
              {
                'text-neutral-600 group-focus-within:text-primary-500 dark:group-focus-within:text-primary-300':
                  !error,
                'text-red-500 group-focus-within:text-red-500': error,
                'after:ml-0.5 after:mr-1 after:text-red-500 after:content-["*"]':
                  required,
                'text-primary-500': isOpen,
              },
              labelClassName
            )}>
            {label}
          </label>
        )}
        <div
          ref={containerRef}
          role='combobox'
          aria-expanded={isOpen}
          aria-haspopup='listbox'
          aria-controls='select-dropdown'
          aria-label={label}
          aria-required={required}
          aria-invalid={error}
          tabIndex={disabled ? -1 : 0}
          onKeyDown={handleKeyDown}
          className='rounded-lg bg-neutral-0 outline-none'>
          <div
            onClick={toggleOpen}
            className={cn(
              'flex w-full cursor-pointer items-center justify-between rounded-lg border border-neutral-10 bg-neutral-0 bg-transparent px-3 py-2 transition-all dark:bg-surface-100',
              disabled
                ? 'cursor-not-allowed opacity-50'
                : 'hover:border-neutral-400',
              {
                'border-neutral-50 focus:border-primary-300 dark:focus:border-primary-300':
                  !error,
                'ring-red-500 focus:ring-red-500': error,
                'border-primary-300 shadow-[0_0_10px_#c9bfff]': isOpen,
              }
            )}>
            <div className='flex items-center gap-2'>
              {selectedOption?.icon && <>{selectedOption.icon}</>}
              <span
                className={cn(
                  'block truncate',
                  !selectedOption ? 'text-neutral-400' : ''
                )}>
                {selectedOption ? selectedOption.label : 'Select option'}
              </span>
            </div>
            <CaretDown
              weight='bold'
              className={cn(
                'h-4 w-4 text-primary-500 transition-transform duration-300',
                isOpen ? 'rotate-180' : '',
                caretClassName
              )}
              aria-hidden='true'
            />
          </div>
        </div>

        {isOpen && (
          <div
            id='select-dropdown'
            ref={dropdownRef}
            role='listbox'
            aria-label={`${label} options`}
            className='absolute z-50 mt-1 max-h-60 w-full overflow-auto rounded-lg border border-neutral-50 bg-neutral-0 py-1 shadow-lg dark:bg-surface-100'>
            {options.map((option, index) => (
              <SelectItem
                key={option.value}
                option={option}
                isSelected={option.value === value}
                isHighlighted={index === highlightedIndex}
                onSelect={() => handleSelect(option.value)}
              />
            ))}
          </div>
        )}

        {error && errorMessage && (
          <p role='alert' className='mt-1 text-sm text-red-500'>
            {errorMessage}
          </p>
        )}
      </div>
    )
  }
)

Select.displayName = 'Select'

export default memo(Select)
