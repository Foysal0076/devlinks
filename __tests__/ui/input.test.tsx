import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Input } from '@/components/ui/Input'

describe('Input component', () => {
  test('renders input with label', () => {
    render(<Input label='Username' id='username' />)
    expect(screen.getByLabelText('Username')).toBeInTheDocument()
  })

  test('renders input with helper text', () => {
    render(<Input helperText='Enter your username' />)
    expect(screen.getByText('Enter your username')).toBeInTheDocument()
  })

  test('displays error message', () => {
    render(<Input error='This field is required' />)
    expect(screen.getByText('This field is required')).toBeInTheDocument()
  })

  test('renders start adornment', () => {
    const startAdornment = { adornment: <span>@</span> }
    render(<Input startAdornment={startAdornment} />)
    expect(screen.getByText('@')).toBeInTheDocument()
  })

  test('renders end adornment and handles click', () => {
    const handleClick = jest.fn()
    const endAdornment = { adornment: <span>Clear</span>, onClick: handleClick }
    render(<Input endAdornment={endAdornment} />)

    const button = screen.getByText('Clear')
    expect(button).toBeInTheDocument()

    fireEvent.click(button)
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  test('applies custom className', () => {
    render(<Input className='custom-class' />)
    expect(screen.getByRole('textbox')).toHaveClass('custom-class')
  })

  test('forwards ref to input element', () => {
    const ref = React.createRef<HTMLInputElement>()
    render(<Input ref={ref} />)
    expect(ref.current).toBeInstanceOf(HTMLInputElement)
  })

  test('matches snapshot', () => {
    const { asFragment } = render(<Input />)
    expect(asFragment()).toMatchSnapshot()
  })
})
