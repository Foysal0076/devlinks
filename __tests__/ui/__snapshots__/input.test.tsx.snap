// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="custom-input group flex flex-col gap-1"
  >
    <div
      class="relative flex flex-col"
    >
      <input
        class="rounded-lg border-none px-4 text-neutral-600 outline-none ring-1 placeholder:text-neutral-100 focus:placeholder:text-primary-100 disabled:bg-neutral-30 disabled:text-neutral-100 dark:bg-surface-100 dark:text-neutral-400 dark:disabled:bg-neutral-400 ring-neutral-50 focus:shadow-[0_0_10px_#c9bfff] focus:ring-primary-300 dark:focus:ring-primary-300"
      />
    </div>
  </div>
</DocumentFragment>
`;
